llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp

```diff
@@ -692,6 +692,11 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
       .customIf(all(typeIsLegalIntOrFPVec(0, IntOrFPVecTys, ST),
                     typeIsLegalIntOrFPVec(1, IntOrFPVecTys, ST)));
 
+  getActionDefinitionsBuilder(G_ATOMICRMW_ADD)
+      .legalFor(ST.hasStdExtA(), {{s32, p0}, {sXLen, p0}})
+      .libcallFor(!ST.hasStdExtA(), {{s8, p0}, {s16, p0}, {s32, p0}, {s64, p0}})
+      .clampScalar(0, s32, sXLen);
```

Comment:
This should fix the i32 selection problem on RV64. All of the SelectionDAG patterns are expecting SXLen so we need to widen to sXLen not s32.

```suggestion
  getActionDefinitionsBuilder(G_ATOMICRMW_ADD)
      .legalFor(ST.hasStdExtA(), {{sXLen, p0}})
      .libcallFor(!ST.hasStdExtA(), {{s8, p0}, {s16, p0}, {s32, p0}, {s64, p0}})
      .clampScalar(0, sXLen, sXLen);
```

Reply:
Thanks for the guidance! I confirmed the SelectionDAG patterns in `RISCVISelLowering.cpp` and have updated the code accordingly, along with refining the test cases.